### ISSUE TITLE
Rule proposal: typescript tuples should use `[]` and arrays `.at()` for accessing members

### ISSUE THREAD

**Issue by nicolassanmar** [typescript-eslint/typescript-eslint#6936]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

There should be a rule that allows enforcing arrays to use `array.at(0)` to access it's members, while allowing tuples to use `tuple[0]` (bracket notation) to do so.
This is useful since using bracket notation on an array of type `T` can return `undefined` if the number is out of range, but the type that typescript returns is only `T`.
Using `.at()` correctly informs the developer about this.

This is not true on tuples, where given a tuple of type `[X, Y]` accessing the parameters with bracket notation should return something of type `X` or `Y`.

Currently there is no rule to enforce this, and I could not find a way to implement this as a custom rule either.

### Fail Cases

```typescript
const array: number[] = [1,2]
array[3] // we should not do this since it could not exist on the element
```

### Pass Cases

```typescript
const tuple: [number,number] = [1,2]
tuple[1] // we know this exists due to the type of the tuple

const array: number[] = [1,2]
array.at(3) // this can return undefined, so the user gets the correct information
```

**Comment by bradzacher** [comment]
The difficult thing about this is that `.at` still only has pretty recent support - less than 2 years in major browsers. Node 14 doesn't support it and it's still LTS.

Additionally there is a TS compiler option which will make array access safe - `noUncheckedIndexAccess`. In fact it'll be safer than `.at` because TS can track the result of checking `arr[I]` and narrow the type - but TS cannot narrow the type of `arr.at(i)` as it is a function call.

[playground example](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&target=99&useUnknownInCatchVariables=true#code/CYUwxgNghgTiAEYD2A7AzgF3rGAuemMAligOYDaAugNwBQ9O5AjJQHQQhkYAWdRAZvAAUjFvACEAXngoArhAgBKeAG9a8bDBjM2HLr1oBfBltZQMQpovadSPPoJGnzlxcrUaNOMxas39RrRAA)

---

That being said - we do have #6401 and the associated PR #6411 which are on the track to doing this style of enforcement.

I'm not sure if we would want to take that the whole way to always using `.at` though. Cc @JoshuaKGoldberg

**Comment by JoshuaKGoldberg** [comment]
Yeah I don't think we should implement this in typescript-eslint core. It's a very particular practice that is mirrored by a TypeScript compiler option - and not (yet?) adopted by the community. 

We (linters) have been hurt in the past by implementing rules that enforce not-seen-as-best-by-everyone(-yet?) best practices not adopted by the community. People get real mad if they think we're trying to enforce some seemingly arbitrary code style on them... This one certainly couldn't be enabled in a recommended rule set, and even if we provide it as an option it's unclear yet whether many users would want it.

I'd love to see it implemented in a shared ESLint plugin using APIs described in [our custom rule docs](https://typescript-eslint.io/custom-rules). Please do! 😄

**Comment by nicolassanmar** [comment]
@bradzacher Thanks! I did not know about `noUncheckedIndexedAccess` and it does solve the issue I have.

@JoshuaKGoldberg I agree with your points. Also, considering that we have the `noUncheckedIndexedAccess`, this rule will not make sense.

Thank you both!
